### DISCUSSION TITLE
advanced mixing options / control busses

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#354] [Ideas]
as discussed in chat: being able to mix haps into certain busses and control those busses would be great.

a maybe unrealistic example:

```js
s('bd').orbit('kick'),
note("a3,c3,g3").sidechain('kick')
```

so the general idea is to be able to use the already existing orbits as inputs to effects.
tidal has https://club.tidalcycles.org/t/tidal-1-7-control-busses-midi-control-input/2934 for that

**Comment by felixroos** [comment]
related idea: it could be a powerful performance optimization if audio nodes would only be created once for haps that have the same whole and the same value for the param in question. example:

```js
note("c2").s("sawtooth,square").cutoff(800)
```

right now, the above snippet will create 4 audio nodes in the first cycle:

- sawtooth oscillator + filter
- square oscillator + filter



one could also say the audio graph is a tree, as branches will never be connected together again.
now if the webaudio output would check all params for matches, the first cycle could be played with only 3 nodes

- sawtooth oscillator
- square oscillator
- filter

where both oscillators are connected to the same filter node. this will only work if the haps have the same whole and the cutoff value is the same..

this also poses a general dsp question: will the two graphs (3 vs 4 nodes) have a different output? in other words: are dsp graphs associative, e.g. `(a+b)*c =?= a*c + b*c` where a and b are the oscillators and c is the filter

**Comment by yaxu** [comment]
If I remember right, for a pattern like `s('bd').cutoffbus(5, saw.range(100,3000).segment(64))`  control busses work by making a trigger message like `{s:'bd', cutoff: 'c5'}` which tells superdirt to expect cutoff values to be sent on control bus number 5. Then the scheduler no longer throws away haps without an onset, but looks for control values to send to the bus (which in the case of superdirt ends up being separate osc messages sent straight to the synth).
